File: cosmoauthorization/src/java/it/csi/cosmo/cosmoauthorization/business/service/FruitoriService.java
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package it.csi.cosmo.cosmoauthorization.business.service;

import java.util.List;
import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaEndpointFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaSchemaAutenticazioneFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaEndpointFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaSchemaAutenticazioneFruitoreRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CredenzialiAutenticazioneFruitore;
import it.csi.cosmo.cosmoauthorization.dto.rest.EndpointFruitore;
import it.csi.cosmo.cosmoauthorization.dto.rest.Fruitore;
import it.csi.cosmo.cosmoauthorization.dto.rest.FruitoriResponse;
import it.csi.cosmo.cosmoauthorization.dto.rest.OperazioneFruitore;
import it.csi.cosmo.cosmoauthorization.dto.rest.SchemaAutenticazioneFruitore;

/**
 *
 */

public interface FruitoriService {

  /**
   * Metodo per eliminare un fruitore dal database
   *
   * @param id e' l'identificativo numerico del fruitore da eliminare
   */
  Fruitore deleteFruitore(Long id);

  /**
   * Metodo che restituisce una lista di fruitori validi. Questa lista ha un numero di elementi pari
   * al valore indicato da filter o, se filter non e' valorizzato, un numero di elementi pari al
   * valore definito nella tabella di configurazione
   *
   * @param filter indica i filtri da utilizzare nella ricerca dei fruitori
   * @return una lista di fruitori validi
   */
  FruitoriResponse getFruitori(String filter);

  /**
   * Metodo che restituisce un fruitore specifico
   *
   * @param apiManangerId del fruitore che deve essere cercato nel database
   * @return fruitore presente nel database
   */
  Fruitore getFruitoreApiManagerId(String apiManangerId);

  /**
   * Metodo che restituisce un fruitore specifico
   *
   * @param id del fruitore che deve essere cercato nel database
   * @return fruitore presente nel database
   */
  Fruitore getFruitore(Long id);

  /**
   * Metodo che inserisce un fruitore sul database
   *
   * @param fruitore e' il fruitore che deve essere inserito sul database
   * @return il fruitore inserito sul database
   */
  Fruitore createFruitore(CreaFruitoreRequest fruitore);

  /**
   * Metodo che aggiorna un fruitore gia' esistente nel database
   *
   * @param fruitore e' il fruitore che deve essere aggiornato
   * @return il fruitore aggiornato nel database
   */
  Fruitore putFruitore(Long id, AggiornaFruitoreRequest fruitore);

  List<OperazioneFruitore> getOperazioniFruitore();

  /**
   * @param idFruitore
   * @param idEndpoint
   */
  void deleteEndpointFruitore(Long idFruitore, Long idEndpoint);

  /**
   * @param idFruitore
   * @param idSchemaAutenticazione
   */
  void deleteSchemaAuthFruitore(Long idFruitore, Long idSchemaAutenticazione);

  /**
   * @param idFruitore
   * @param body
   * @return
   */
  EndpointFruitore postEndpointsFruitori(Long idFruitore, CreaEndpointFruitoreRequest body);

  /**
   * @param idFruitore
   * @param body
   * @return
   */
  SchemaAutenticazioneFruitore postSchemiAuthFruitori(Long idFruitore,
      CreaSchemaAutenticazioneFruitoreRequest body);

  /**
   * @param idFruitore
   * @param idEndpoint
   * @param body
   * @return
   */
  EndpointFruitore putEndpointFruitore(Long idFruitore, Long idEndpoint,
      AggiornaEndpointFruitoreRequest body);

  /**
   * @param idFruitore
   * @param idSchemaAutenticazione
   * @param body
   * @return
   */
  SchemaAutenticazioneFruitore putSchemaAuthFruitore(Long idFruitore, Long idSchemaAutenticazione,
      AggiornaSchemaAutenticazioneFruitoreRequest body);

  /**
   * @param body
   * @return
   */
  Fruitore postFruitoriAutentica(CredenzialiAutenticazioneFruitore body);

}

File: cosmowcl/src/app/shared/services/fruitori.service.ts
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
import {
  HttpClient,
  HttpParams,
} from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';

import {
  AggiornaEndpointFruitoreRequest,
} from '../models/api/cosmoauthorization/aggiornaEndpointFruitoreRequest';
import {
  AggiornaFruitoreRequest,
} from '../models/api/cosmoauthorization/aggiornaFruitoreRequest';
import {
  AggiornaSchemaAutenticazioneFruitoreRequest,
} from '../models/api/cosmoauthorization/aggiornaSchemaAutenticazioneFruitoreRequest';
import {
  AutorizzazioniFruitoreResponse,
} from '../models/api/cosmoauthorization/autorizzazioniFruitoreResponse';
import {
  CreaEndpointFruitoreRequest,
} from '../models/api/cosmoauthorization/creaEndpointFruitoreRequest';
import {
  CreaFruitoreRequest,
} from '../models/api/cosmoauthorization/creaFruitoreRequest';
import {
  CreaSchemaAutenticazioneFruitoreRequest,
} from '../models/api/cosmoauthorization/creaSchemaAutenticazioneFruitoreRequest';
import {
  EndpointFruitore,
} from '../models/api/cosmoauthorization/endpointFruitore';
import { Fruitore } from '../models/api/cosmoauthorization/fruitore';
import {
  FruitoriResponse,
} from '../models/api/cosmoauthorization/fruitoriResponse';
import {
  OperazioneFruitore,
} from '../models/api/cosmoauthorization/operazioneFruitore';
import {
  SchemaAutenticazioneFruitore,
} from '../models/api/cosmoauthorization/schemaAutenticazioneFruitore';
import { ApiUrls } from '../utilities/apiurls';
import { Utils } from '../utilities/utilities';

@Injectable()
export class FruitoriService {

  constructor(
    private http: HttpClient
  ) {
    // NOP
  }

  /** GET: Restituisce un fruitore. */
  getFruitore(id: number): Observable<Fruitore> {
    Utils.require(id);

    return this.http.get<Fruitore>(ApiUrls.GET_FRUITORE.replace('{id}', id.toString()));
  }

  /** GET: Restituisce una lista di fruitori. */
  getFruitori(filter: string): Observable<FruitoriResponse> {
    const options = filter ?
      { params: new HttpParams().set('filter', filter) } : {};
    return this.http.get<FruitoriResponse>(ApiUrls.FRUITORI, options);
  }

  /** GET: Restituisce una lista di fruitori. */
  getAutorizzazioniFruitori(filter: string): Observable<AutorizzazioniFruitoreResponse> {
    const options = filter ?
      { params: new HttpParams().set('filter', filter) } : {};
    return this.http.get<AutorizzazioniFruitoreResponse>(ApiUrls.AUTORIZZAZIONI_FRUITORE, options);
  }

  /** DELETE: Elimina un fruitore. */
  deleteFruitore(id: number): Observable<void> {
    Utils.require(id);

    return this.http.delete<void>(ApiUrls.GET_FRUITORE.replace('{id}', id.toString()));
  }

  /**
   * POST: Salva un Fruitore.
   *
   * @param fruitore Fruitore da salvare.
   * @returns Observable<Fruitore> il fruitore salvato.
   */
  creaFruitore(request: CreaFruitoreRequest): Observable<Fruitore> {
    Utils.require(request);

    return this.http.post<Fruitore>(ApiUrls.FRUITORI, request);
  }

  /**
   * PUT: Aggiorna un fruitore.
   *
   * @param fruitore Fruitore da aggiornare.
   * @returns Observable<Fruitore> l'fruitore aggiornato.
   */

  aggiornaFruitore(id: number, request: AggiornaFruitoreRequest): Observable<Fruitore> {
    Utils.require(id);
    Utils.require(request);

    return this.http.put<Fruitore>(ApiUrls.GET_FRUITORE.replace('{id}', id.toString()), request);
  }

  creaSchemaAutenticazioneFruitore(idFruitore: number, request: CreaSchemaAutenticazioneFruitoreRequest):
    Observable<SchemaAutenticazioneFruitore>
  {
    Utils.require(idFruitore, 'idFruitore');

    return this.http.post<SchemaAutenticazioneFruitore>(
      ApiUrls.SCHEMI_AUTENTICAZIONE_FRUITORE
        .replace('{idFruitore}', idFruitore.toString()),
      request
    );
  }

  aggiornaSchemaAutenticazioneFruitore(idFruitore: number, idSchema: number, request: AggiornaSchemaAutenticazioneFruitoreRequest):
    Observable<SchemaAutenticazioneFruitore>
  {
    Utils.require(idFruitore, 'idFruitore');
    Utils.require(idSchema, 'idSchema');
    Utils.require(request);

    return this.http.put<SchemaAutenticazioneFruitore>(
      ApiUrls.SCHEMA_AUTENTICAZIONE_FRUITORE
        .replace('{idFruitore}', idFruitore.toString())
        .replace('{idSchema}', idSchema.toString()),
      request
    );
  }

  eliminaSchemaAutenticazioneFruitore(idFruitore: number, idSchema: number): Observable<void> {
    Utils.require(idFruitore, 'idFruitore');
    Utils.require(idSchema, 'idSchema');

    return this.http.delete<void>(
      ApiUrls.SCHEMA_AUTENTICAZIONE_FRUITORE
        .replace('{idFruitore}', idFruitore.toString())
        .replace('{idSchema}', idSchema.toString())
    );
  }

  testSchemaAutenticazioneFruitore(idSchema: number): Observable<object> {
    Utils.require(idSchema, 'idSchema');

    return this.http.post<object>(
      ApiUrls.TEST_SCHEMA_AUTENTICAZIONE_FRUITORE
        .replace('{idSchema}', idSchema.toString()), {}
    );
  }

  creaEndpointFruitore(idFruitore: number, request: CreaEndpointFruitoreRequest): Observable<EndpointFruitore> {
    Utils.require(idFruitore, 'idFruitore');

    return this.http.post<EndpointFruitore>(
      ApiUrls.ENDPOINTS_FRUITORE
        .replace('{idFruitore}', idFruitore.toString()),
      request
    );
  }

  aggiornaEndpointFruitore(idFruitore: number, idEndpoint: number, request: AggiornaEndpointFruitoreRequest): Observable<EndpointFruitore> {
    Utils.require(idFruitore, 'idFruitore');
    Utils.require(idEndpoint, 'idEndpoint');
    Utils.require(request);

    return this.http.put<EndpointFruitore>(
      ApiUrls.ENDPOINT_FRUITORE
        .replace('{idFruitore}', idFruitore.toString())
        .replace('{idEndpoint}', idEndpoint.toString()),
      request
    );
  }

  eliminaEndpointFruitore(idFruitore: number, idEndpoint: number): Observable<void> {
    Utils.require(idFruitore, 'idFruitore');
    Utils.require(idEndpoint, 'idEndpoint');

    return this.http.delete<void>(
      ApiUrls.ENDPOINT_FRUITORE
        .replace('{idFruitore}', idFruitore.toString())
        .replace('{idEndpoint}', idEndpoint.toString())
    );
  }

  getOperazioniFruitore(): Observable<OperazioneFruitore[]> {
    return this.http.get<OperazioneFruitore[]>(ApiUrls.OPERAZIONI_FRUITORE);
  }
}
